lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -1187,69 +1187,83 @@ void LocalEnforcer::add_rules_for_unsuspended_credit(
       session->get_config(), rules_to_add, RulesToProcess{}, false);
 }
 
+bool LocalEnforcer::handle_credit_update_failure(
+    const CreditUpdateResponse& credit_update_resp,
+    UpdateChargingCreditActions* actions) const {
+  const std::string& imsi       = credit_update_resp.sid();
+  const std::string& session_id = credit_update_resp.session_id();
+  const uint32_t& ckey          = credit_update_resp.charging_key();
+  const uint32_t& result_code   = credit_update_resp.result_code();
+
+  if (credit_update_resp.success()) {
+    return true;
+  }
+  // handle permanent failure -> terminate
+  if (DiameterCodeHandler::is_permanent_failure(result_code)) {
+    MLOG(MERROR) << session_id << " Received permanent failure result code "
+                 << result_code << " during update" << session_id;
+    actions->sessions_to_terminate.insert(ImsiAndSessionID(imsi, session_id));
+    return false;
+  } else if (DiameterCodeHandler::is_transient_failure(result_code)) {
+    // handle transient failure -> suspend
+    MLOG(MERROR) << session_id << " Received transient failure result code "
+                 << result_code << " during update" << session_id;
+    actions->suspended_credits.insert(
+        ImsiSessionIDAndCreditkey{imsi, session_id, ckey});
+  } else {
+    // TODO( ): deal with other error codes
+    MLOG(MERROR) << "Received Unimplemented result code " << result_code
+                 << " for " << session_id << " during update. Not action taken";
+  }
+  return true;
+}
```

Comment:
I remember I tried with a similar function and I felt that the function was mixing too many things, doing stuff in some objects, and returning a value. That is why I prefered to had the logic on `update_charging_credits`

Reply:
Yeah it's a bit opaque but tried to make it so that the boolean value returned just indicates whether we want to further process the credit or not. 
The `update_charging_credits` function was getting too long/ hard to read :/

Reply:
I can add a comment on the line where we call the function too, if that helps with clarity.

Reply:
yes, maybe a comment will help us in the future.

Reply:
will do!